Packages/nadena.dev.modular-avatar/Runtime/ModularAvatarObjectRemover.cs

```diff
@@ -0,0 +1,133 @@
+        public List<GameObject> objectsToRemove = new List<GameObject>();
+        private List<GameObject> _lastObjectsToRemove;
+
+        private void OnEnable()
+        {
+            objectsToRemove?.RemoveAll(x => !x);
+            OnListChange();
+        }
+
+        private new void OnDestroy()
+        {
+            if (Application.isPlaying) return;
+            hideInHierarchy = false;
+            removePrefabComponents = false;
+            Unhide();
+            RestorePrefabComponents();
+            objectsToRemove.Clear();
+            _lastObjectsToRemove = null;
+        }
+
+        public void OnListChange()
+        {
+            objectsToRemove.RemoveAll(x => !x);
+            if (Application.isPlaying) return;
+            if (_lastObjectsToRemove != null)
+            {
+                foreach (var newValue in objectsToRemove) _lastObjectsToRemove.Remove(newValue);
+                foreach (var removedFromList in _lastObjectsToRemove)
+                {
+                    if (!removedFromList) continue;
+                    removedFromList.hideFlags &= ~HideFlags.HideInHierarchy;
+                    foreach (var removedComponent in CollectComponentsToRestore(new List<RemovedComponent>(), removedFromList.transform))
+                    {
+                        removedComponent.Revert();
+                    }
+                }
+            }
+
+            _lastObjectsToRemove = new List<GameObject>(objectsToRemove);
+            if (hideInHierarchy) Hide();
+            if (removePrefabComponents) RemovePrefabComponents();
+            if (keepDisabled)
+            {
+                foreach (var obj in objectsToRemove) obj.SetActive(false);
+            }
+        }
+
+        private void CollectComponentsToRemove(List<Component> list, Transform current)
+        {
+            list.AddRange(current.GetComponents(typeof(Component)).Where(PrefabUtility.IsPartOfAnyPrefab));
+            foreach (Transform child in current)
+            {
+                CollectComponentsToRemove(list, child);
+            }
+        }
+
+        private List<RemovedComponent> CollectComponentsToRestore(List<RemovedComponent> list, Transform current)
+        {
+            if (!PrefabUtility.IsPartOfAnyPrefab(current.gameObject)) return list;
+            list.AddRange(PrefabUtility.GetRemovedComponents(current.gameObject));
+            foreach (Transform child in current)
+            {
+                CollectComponentsToRestore(list, child);
+            }
+
+            return list;
+        }
+
+        public void RemovePrefabComponents()
+        {
+            var toDelete = new List<Component>();
+
+            foreach (var toRemove in objectsToRemove)
+            {
+                if (removePrefabComponents) CollectComponentsToRemove(toDelete, toRemove.transform);
+            }
+
+            // NOTE: this is non-destructive, all components from prefabs can be restored with no data loss
```

Comment:
We do lose information on any components that were already removed on the prefab instance before we applied this change. This in turn means that undo won't work properly, as we don't record these removals in the undo record and don't retain the information needed to _not_ restore components that were already removed.

Reply:
Kinda same as above, at this point I think its all out of scope of this idea, whole idea is to have some easy way of getting rid of parts of prefab you dont want to use without need of breaking the prefab and ability to restore it in case you change mind. Allowing for quicker creation, better optimisation and cleaner project, so I don't see any use cases for such situations.
For needs of this repo I guess removing feature can be removed, as its not too important anyways, but the idea was to just have nice way to automate removing all the components from all the stuff you don't use, so if restoring will restore too much its not a big deal compared to saving time on removing and searching for it manually without breaking anything